### PR TITLE
GitHub Sync: Get patch using local Git

### PR DESCRIPTION
**Pull request by val-fed**

### CODE REVIEW

File: common/repo/GitRepo.java [hunk 1]
```diff
@@ -461,6 +461,11 @@ public boolean hasChanges(String branch) {
     return (getCommits(branch).size() > 1) || (!getUncommittedFiles().isEmpty());
   }
 
+  @Override
+  public String getPatch(String commitOrBranch, String filename) {
```
Comment (oferb) [inline common/repo/GitRepo.java]: s/commitOrBranch/referenceCommitOrBranch
According to the docs, this is for "changes you have in your working tree relative to the named <commit>".
https://git-scm.com/docs/git-diff

File: common/repo/Repo.java [hunk 2]
```diff
@@ -81,5 +81,10 @@
   String getRemoteURL();
   /** Checks if there are commits added since master, or any uncommitted files */
   boolean hasChanges(String branch);
+  /**
+   * Gets patch(diff) between current file and the file of the base branch(usually "master"). It's
```
Comment (oferb) [inline common/repo/Repo.java]: See previous comment - isn't this between the current file and the file in the passed `commitOrBranch`?

File: common/repo/repo.proto [hunk 3]
```diff
@@ -44,6 +44,9 @@ message File {
   // Empty if referring to a local file
   string commit_id = 5;
   Action action = 6;
+  // Git patch(diff) between current file and the file of the base
```
Comment (oferb) [inline common/repo/repo.proto]: Why do we need this new field?
If the repo is available to the sync job, we can always fetch it from the repo.

File: common/repo/Repo.java (hunk 2 above)
Comment (val-fed) [reply common/repo/Repo.java]: Done.

File: common/repo/Repo.java [hunk 4]
```diff
@@ -81,5 +81,12 @@
   String getRemoteURL();
   /** Checks if there are commits added since master, or any uncommitted files */
   boolean hasChanges(String branch);
+  /**
+   * Gets patch(diff) between current file and the file in the passed `referenceCommitOrBranch`.
+   * It's possible to use branch name or commit ID. The response can have one and more diff hunks
+   */
+  String getPatch(String referenceCommitOrBranch, String filename);
+
+  String getTheLatestCommitIdOfBranch(String branch);
```
Comment (oferb) [inline common/repo/Repo.java]: How about
`/getTheLatestCommitIdOfBranch/getMostRecentCommit`

File: tools/reviewer/job/sync/LineNumberConverter.java [hunk 5]
```diff
@@ -164,7 +164,21 @@ private static int processDiffPatch(
       }
     } else {
       if (side.equals(Side.LEFT)) {
-        return lineNumberToPositionLeftSide.get(number);
+        if (lineNumberToPositionLeftSide.containsKey(number)) {
+          return lineNumberToPositionLeftSide.get(number);
+          // It isn't possible to convert Reviewer's line number to GitHub patch(diff)'s position if
+          // the patch doesn't contain this line.
+        } else {
+          // XXX 1)How to continue the sync without exception.
```
Comment (oferb) [inline tools/reviewer/job/sync/LineNumberConverter.java]: I would choose any line (maybe the closest, if that's possible), and add to the comment:
"Synced from line X"

File: common/repo/Repo.java (hunk 4 above)
Comment (val-fed) [reply common/repo/Repo.java]: Done.

File: tools/reviewer/job/sync/LineNumberConverter.java (hunk 5 above)
Comment (val-fed) [reply tools/reviewer/job/sync/LineNumberConverter.java]: Added TODO for it.

**Pull request merged**